solve changes

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed



#### Other comments

When a single equation has several factors, the factors are now removed as they are found. This makes the solution process more efficient, e.g.
```python
>>> t=time();ok=solve(expand(Mul(*[(x-i) for i in var('a:f')])),var('a:f'));time()-t
3.4788060188293457
```
In master that takes 14.5 seconds.

It is not that the expression is factored -- that case is already handled by `solve`. In this case the expanded expression is linear in each of the parameters when viewed individually so the linear solution can be found easily and then the expression (which contains that solution as a linear factor, though not explicitly) can have that factor removed.
```python
>>> eq= expand((x-a)*(x-b)); eq
a*b - a*x - b*x + x**2
>>> eq= expand((x-a)*(x-b)); eq
a*b - a*x - b*x + x**2
>>> solve(eq,a)
[x]
>>> div(eq,a-x)
(b - x, 0)
```

The docstring has been updated to indicate that the automatic solving for undetermined coefficients must be enabled.

The implicit assumption of `solve_undetermined_coeffs` is that the system is linear in specified parameters. This is now enforced. This means that systems formerly recognized and solved as coefficient systems may no longer be recognized as such:

***formerly***
```python
>>> solve(a**2*x+b-4*x+3, a, b)
[(-2, -3), (2, -3)]
>>> solve(a**2*x+b-4*x+3, {a,b})  # /!\ ambiguously ordered solution
[(-2, -3), (2, -3)]
```
***now***
```python
>>> solve(a**2*x+b-4*x+3,a,b)
[(a, -a**2*x + 4*x - 3)]
>>> solve(a**2*x+b-4*x+3,{a,b})
[{b: -a**2*x + 4*x - 3}]
```

For those that like explicit solutions showing every symbol requested -- even if it is an arbitrary parameter -- the flag `tuple=True` can be used:
```python
>>> solve(x, x, tuple=True)
[(0,)]
>>> solve(x, x, y, z, tuple=True)
[(0, y, z)]
>>> solve([x, y], x, y, z, tuple=True)
[(0, 0, z)]
```
(This was formerly only used for lists of equations which were nonlinear in some way.)

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * solve can return full tuple(s) containing values for all symbols by using ``tuple=True``; set and dict output is ecomonical and does not report symbols that can have arbitrary values
  * solve returns dict when solving single equation for one of several provided symbols (was formerly tuple)
  * more efficient solving of an equation for several factors is implemented
  * when passing a single polynomial expression in some variable `x` and desiring to find solutions for symbols that appear linearly as symbolic coefficients, use the flag `match=True`, e.g. `solve(Eq(a*x**2 + b*x,  3*x**2), a, b, match=True) -> {a: 3, b: 0}` otherwise a more general algebraic solution like `[{a: -b/x + 3}]` will be returned.
  * `solve_undetermined_coeffs` now enforces that the coefficient system is linear in parameters; systems that are nonlinear in parameters will have to be created explicitly and passed to `solve` for a solution.
* other
  * added a "solve_output" explanation file to the documentation
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -136,15 +137,15 @@ def test_solve_args():
-    # multiple symbols might represent an undetermined coefficients system
+    # multiple symbols: return first solution (linear if possible)
+    assert solve(x + y - 3, [x, y]) == [{x: 3 - y}]
+    # multiple symbols might represent a coefficients system
+    # - that is linear
     assert solve(a + b*x - 2, [a, b]) == {a: 2, b: 0}
+    assert solve((a*x + b*x - b + d), (a, b)) == {a: -d, b: d}
```
I don't think that `solve` should do this. This is conceptually very different from what `solve` usually does. These results do not hold for almost all values of `x` e.g.:
```python
In [18]: eq = a + b*x - 2

In [19]: solve(eq, [a, b], dict=True)
Out[19]: [{a: 2, b: 0}]

In [20]: solve(eq.subs(x, 1), [a, b], dict=True)
Out[20]: [{a: 2 - b}]

In [21]: solve(eq.subs(x, 2), [a, b], dict=True)
Out[21]: [{a: 2 - 2⋅b}]
```
For any particular value of `x` there are infinitely many solutions. The correct return here would be:
```python
solve(eq.subs(x, 2), dict=True)
[{a: 2 - x⋅b}]
```

smichr replied:
Without sentiment of defense, I note that the result in master for your example is the same and it has been like this since 2011 (2ff42e6893812f42147d8cff16dc3f371021ab63). If the user wants to solve for `a` or `b` then they can give a single symbol or not give any symbols and indicate what they don't want (e.g., using `exclude=[x]`).

I actually had in mind comments you have made (but I can't find now) about results which are nearly obviously not what the user would have wanted. (If you can find that thread, it would be helpful).

It might be helpful to separate out the `solve_undetermined_coeffs` from `solve`...but that will be a regression for folks for whom such a request to solve makes sense.

oscarbenjamin replied:
I can see that it is already the case on master for that example but I consider that as bug to be fixed and I certainly don't want to add more cases like that.

When the user asks to solve a system of equations with certain symbols given as unknowns any other symbols should be understood as parameters rather than unknowns. It isn't for solve to decide that x=0 if x is not an unknown. We need solve to return a result that is valid generically for almost all possible values of x.

It's quite clear here what that result should be:
```python
In [22]: solve([eq], [a, b])
Out[22]: {a: -b⋅x + 2}

In [23]: linsolve([eq], [a, b])
Out[23]: {(-b⋅x + 2, b)}

In [24]: nonlinsolve([eq], [a, b])
Out[24]: {(-b⋅x + 2, b)}
```
The special case that gives this result should just be removed:
```python
In [25]: solve(eq, [a, b])
Out[25]: {a: 2, b: 0}
```
The equation is underdetermined so solve should not return something that indicates a unique solution.

smichr replied:
Note that this result is for the one-equation-many-symbols case. If we don't treat that differently than n-equations-more-than-n-symbols then do you propose that we raise an error when more than one symbol is provided? Or just pick an arbitrary variable for which to solve? What's wrong with
```python
eq = f(a,b,x)
solve(eq) -> solve returns a solution for a, b or x
solve(eq, exclude=[x]) or solve([eq],(a,b)) -> solve solves for a or b
solve(eq, (a,b)) -> solve gives values of a AND b that make the equation 0 (case of undetermined coefficients)
```
The case of solving for undetermined coefficients is not an uncommon request. Do you agree that splitting off that behavior is a regression? Any ideas on how to accomplish that?

ping @asmeurer

oscarbenjamin replied:
The expected output is this:
```python
In [10]: a, b, x = symbols('a, b, x')

In [11]: eq = a + b*x - 2

In [12]: solve([eq], [a, b])
Out[12]: {a: -b⋅x + 2}
```
This should give the same result:
```python
In [13]: solve(eq, [a, b])
Out[13]: {a: 2, b: 0}
```
There should be no distinction between `eq` and `[eq]` here and the result for `[eq]` is the correct one in this case. It is arbitrary whether `a` should be given in terms of `b` or vice versa in general but in this case it's better to give `a` in terms of `b` because it doesn't require dividing by `x` which might be zero.

oscarbenjamin replied:
I would consider fixing it so that `eq` and `[eq]` return the same (correct) results to be a bugfix rather than a regression.

smichr replied:
cf docstring for solve which explicitly states that one equation with more than one variable will attempt a "undetermined coefficients" solution:
```
Single expression and more than one symbol:
        * When there is a linear solution:
            >>> solve(x - y**2, x, y)
            [(y**2, y)]
            >>> solve(x**2 - y, x, y)
            [(x, x**2)]
            >>> solve(x**2 - y, x, y, dict=True)
            [{y: x**2}]
        * When undetermined coefficients are identified:
            * That are linear:
                >>> solve((a + b)*x - b + 2, a, b)
                {a: -2, b: 2}
            * That are nonlinear:
                >>> solve((a + b)*x - b**2 + 2, a, b, set=True)
                ([a, b], {(-sqrt(2), sqrt(2)), (sqrt(2), -sqrt(2))})
```
Should this be raise on the mailing list before making a change to this?

oscarbenjamin commented:
> * symbols used internally to replace non-symbols are now assigned so symbols reflect ordering of non-symbols
> 
> * a helper `coefficient_system` was added to identify coefficients of generators containing a common symbol
> 
> * solution dictionaries are now sorted to make it easier to visually locate variables of interest
> 
> * finding solutions of a single equation for several variables is more efficient

Can you separate these into different PRs? It's very difficult to review the changes for each of these separately when they are all jumbled together in a single PR.

smichr commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1460,12 +1459,12 @@ def test_issue_5901():
     assert solve([f(x) - 3*f(x).diff(x)], f(x)) == \
         {f(x): 3*D}
     assert solve([f(x) - 3*f(x).diff(x), f(x)**2 - y + 4], f(x), y) == \
-        [{f(x): 3*D, y: 9*D**2 + 4}]
```
This should be a dict because even though it is non-linear in `f(x)`, the solution is obtained by linear substitution (solving first for linear `f(x)` and substituting into second to solve for linear `y`).

smichr replied:
But that is a byproduct piece of information and we are trying to move away from multiple forms of output. So I think this is ok. The symbols were ordered and a system was passed, so a tuple of ordered solutions is returned.

smichr commented:
Let's see what breaks when removing undetermined coefficients detection in solve

smichr commented:
I'll handle removal of undetermined coefficients solving in another PR.

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1460,12 +1459,12 @@ def test_issue_5901():
         {f(x): 3*D}
     assert solve([f(x) - 3*f(x).diff(x), f(x)**2 - y + 4], f(x), y) == \
-        [{f(x): 3*D, y: 9*D**2 + 4}]
+        [(3*D, 9*D**2 + 4)]
```
Why has this changed?

I don't really understand how this is decided:
```
In [1]: solve([x, y], x, y)
Out[1]: {x: 0, y: 0}

In [2]: solve([x**2, y], x, y)
Out[2]: {x: 0, y: 0}

In [3]: solve([x**2-1, y], x, y)
Out[3]: [(-1, 0), (1, 0)]

```

smichr replied:
If a system has a single solution it comes back as a dict, otherwise as a list of tuples *if solutions for all symbols have been found*.
```python
>>> solve([x**2-1, y])  # can't return tuple or else result is ambiguous
[{x: -1, y: 0}, {x: 1, y: 0}]
```
That's only partially true, however, as this example shows. In the case of `x**2, y` the system is split into two and the final result is a single solution so the splitter code returns the dict
```
            # Return a list with one dict as just the dict
            if len(sols) == 1:
                return sols[0]
            return sols
```
There is no communication that the single equation `[x**2]` was non-linear, only that it had a single solution.

I the case of the changed code, since the system was coupled and non-linear, the solution came back in tuple form indicating that the system had, potentially, more than one solution. Since a dictionary result was not requested, a list of dictionaries should not be returned. 

Current behavior should be tuple of solutions or single dictionary if symbols are provided, else single dictionary or list of dictionaries if symbols are not provided.

oscarbenjamin replied:
I honestly can't wrap my head around what the correct return value should be.

How difficult/breaking would it be to change solve so that the type of the return value is determined purely by the types of the arguments?

smichr replied:
This was discussed at https://groups.google.com/g/sympy/c/v_YLkX4QuRY (that was the thread I was looking for earlier)

I think current behavior should be

- single equation, single variable: list of expressions
- single equation, 0 or more than 1 variable: list of dicts (*)
- list of equation(s), single solution -> dict
- list of equation(s), multiple solutions, ordered symbols with solutions for each: list of tuples
- list of equation(s), multiple solutions, unordered symbols or fewer solutions than symbols: list of dicts

(*) handling undetermined coefficients throws the wrench into this, however, because it will return a dict. But if that is removed, then the above scheme should hold.

And if one doesn't want that, select `dict=True` to always get a list of dictionaries.

oscarbenjamin replied:
So isn't this case 3?

smichr replied:
Yes. I think we should minimize the dictionary return type, however. This is akin to returning a single expression instead of a singleton list if a single solution to one equation is found. So if we can limit the bare dictionary to a fully linear system that would be better. In that case, case 3 should read "list of *linear* equations" and give dict or [] (if there is no solution).

oscarbenjamin replied:
Honestly none of this makes any sense to me at all. I can't imagine that it makes much sense to anyone else either but I assume that people will have written code that depends on the particular return types that they saw when calling solve so that any change now would break something for someone.

I'm just trying to work out if there's any way to make this more predictable or consistent without breaking too much. The most problematic cases are where the users passes exactly the same type of inputs but gets different types of outputs depending the value of the inputs. Distinguishing between linear and nonlinear is precisely the kind of thing that we should not do here.

It should be possible without looking at the equations themselves to say what the form of the output will be. In this case we have a list of two equations and two unknowns given explicitly. Can there be a unique rule that describes what the return type should be when there are two equations with two unknowns?

smichr replied:
Least breaking, I think, would be two types. A single equation and single variable: list of expressions; multiple equations or multiple symbols, list of dictionaries. But retain existing option to request output of dict=True to always give a list of dicts. Anything else can be handled at the user level to convert to other types.

oscarbenjamin replied:
I think I've come to the conclusion that it's best not to change anything until there is a fully worked out new API. It could be called `solve2` or something but until we have that I think that the return types from `solve` should just be left alone.

smichr replied:
But what about `solve(eqs, syms)` where `eqs=[x - 1]` and `syms=[x]` or `solve(eqs, x)`? It's probably easiest to determine by inspection what the result is going to be by considering the variable. Hopefully by name or close context it is clear if one is dealing with a symbol or a container of symbols. The former could give the list of expressions as the result and the other, the list of dictionaries regardless of how the equation is passed. So the rule could be: 1) variable is a symbol -> list of expressions; 2) symbol(s) passed as a list -> list of dictionaries.

smichr replied:
> not to change anything

In the case of the changed doctest this should be considered wrong. It (and the correction) are inconsistent with the existing behavior in the list 1-5 above.

smichr replied:
Here is the current logic for variables passed as symbols:
```
if equation is a boolean expression or a list contains a boolean
	boolean value or expression
		>>> solve((x**2-4, x>1))
		Eq(x, 2)
		>>> solve((x**2-4, x>2))
		False

else expression or list of expression(s) is passed
    and there is no solution
		return []
	else
		equation is single expression
			0 symbols
				univariate - list of values
				else - list of dictionaries
			1 symbol
				list of values
			> 1 symbol
				list of tuples
		list of 1 or more expressions for equation with
			0 symbols
				linear in all variables - dict
				else - list of dictionaries
			1 symbol
				single solution - dict
				else - list of tuples
			> 1 symbol
				linear system and #eqs = #symbols - dict
				else - list of tuples
```

asmeurer replied:
If you wanted to flesh this out, this would make a good document for the "explanations" section of the documentation.

smichr replied:
Can you point me to a document or do you mean solve docstring, @asmeurer

smichr replied:
Got it: [doc/src/explanation/index.rst](https://github.com/sympy/sympy/pull/23746/files#diff-7bd55f350b73709e128b7a14c1dcc87f1a15685ca61eda91deb5616f4c498855)

asmeurer replied:
Yes that. Look at my improved descriptions of the sections at https://github.com/sympy/sympy/pull/23669. The "explanations" section is where we can go into depth to document things like implementation details and discuss design decisions.

asmeurer replied:
And also see https://diataxis.fr/

smichr replied:
> And also see

NICE! Glad to see that you are using that as a guide.

smichr replied:
When #23669 and this PR are merged I will begin deprecation of `solve` trying to do undetermined coefficients and add the explanation text similar to what I posted at https://groups.google.com/g/sympy/c/v_YLkX4QuRY .

smichr replied:
If there is no objection, this can go in on Monday.

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -523,10 +524,8 @@ def solve(f, *symbols, **flags):
 
             >>> solve(x - y**2, x, y)
-            [(y**2, y)]
+            [{x: y**2}]
```
this is inconsistent with current behavior

oscarbenjamin replied:
Which of the two possible outputs is inconsistent?

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -549,7 +548,7 @@ def solve(f, *symbols, **flags):
             >>> solve(x**2 - y**2/exp(x), x, y, dict=True)
             [{x: 2*LambertW(-y/2)}, {x: 2*LambertW(y/2)}]
             >>> solve(x**2 - y**2/exp(x), y, x)
-            [(-x*sqrt(exp(x)), x), (x*sqrt(exp(x)), x)]
```
inconsistent with current behavior

smichr committed:
enforce linear assumption; handle solve flags

smichr committed:
more efficient solving of 1 eq for many syms

smichr committed:
order dicts; improve tracking of solve i/o

smichr committed:
tests and docstrings updated

smichr committed:
change error to returning None

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -513,31 +529,53 @@ def solve(f, *symbols, **flags):
+        But that might not be what you wanted. To disable that behavior,
+        set the flag ``undetermined=False``. In the future, the only
+        way to get this behavior will be to call directly the currently
+        available function, `solve_undetermined_coeffs`.
```
This could be ma de a sphinx reference so that it generates a clickable link in the html docs.

oscarbenjamin replied:
I guess also in the future the `undetermined` flag will become a no-op.

oscarbenjamin commented:
I think this looks good.

oscarbenjamin commented:
The release note `to disable automatic recognition of undetermined coefficients by solve, use undetermined=True` can be expanded upon. This should explain what this aspect of solve does and that it will change in future (if that is the agreed plan). We should also be clear about whether `undetermined=True` is a feature to be supported indefinitely or just a temporary way to preserve compatibility.

smichr committed:
update solve_undetermined_coeffs

An explicit check that the system is linear in the parameters
is now made and a call to `solve` is avoided if no flags were
provided since the process of collecting and checking the
coefficient system provides the information needed to solve
with `linsolve`.

smichr committed:
add match (default False) flag to solve

It has been debated whether this is a bug fix or a
feature change. Since solving an equation as a set
of equations derived by matching coefficients on
powers of a variable is not the normal use of `solve`
it is being treated as a bug fix in this commit.

For sake of those that might be depending on this
behavior, however, the use of keyword `match=True`
will restore the behavior rather then having to
rewrite the call to `solve` with a call to `solve_undetermined_coeffs`,
e.g.

    solve(a*x + b - 2*x, a, b) to
    solve_undetermined_coeffs(a*x + b - 2*x, [a, b], x)
    instead of
    solve(a*x + b - 2*x, a, b, match=True)

smichr committed:
add solve_output explanation

smichr committed:
rework solve_undetermined_coeffs

smichr committed:
favor canonical linear solutions

First, do not let a single equation be subject to
expansion -- handle the factors first.

Second, while removing linear factors, allow a
rescan of previously attempted symbols since they
may now appear linearly after removing another factor.

smichr committed:
fix swap_back test

smichr committed:
add has_free_arg: faster; no subexpr match

has_free will now attempt a fast literal check
before beginning the slower sub-expr check.

We avoid filtering or checking the patterns since
they may be much larger than the expression to
be tested.

smichr committed:
linear_coeffs has nonstrict option

smichr committed:
rebuild without evaluation

smichr committed:
do not allow cancellation of nonlinears

smichr committed:
Update test_solveset.py

smichr committed:
watch for Eq issues

_lin_eq2dict will handle Eq, but if we exit early from linear_coeffs we have to return the evaluated eq.lhs-eq.rhs; and in _lin_eq2dic we have to process as unevaluated to flag nonlinearity, but filter result in case linear terms cancelled and produced a value of 0 for a key.

So these routines should only give an unevaluated value for a key if the original expression was unevaluated.

smichr committed:
docstring cautions/edits

smichr committed:
syntax fix

smichr committed:
remove use of has; assign directly from keys in d

smichr committed:
fix imports

smichr committed:
efficient linear_eq_to_matrix;strict flags

smichr committed:
use temp flag to keep from expanding

smichr committed:
top import

smichr committed:
more helpful error message if strict; top import

smichr committed:
Update solveset.py

smichr committed:
allow linear_eq_to_dict to return matrices as DM

Also print error info without diagnosis

smichr committed:
Apply suggestions from code review

smichr committed:
Update sympy/solvers/solveset.py

smichr committed:
linsolve gets strict flag, too

smichr committed:
linsolve uses strict=False logic

This is the deprecated logic of linsolve (because expansion will be used to remove singularities if possible).

smichr committed:
pass through strict and _expand

smichr committed:
do not report failing expressions

smichr committed:
SparseMatrix recognizes DoD format

smichr committed:
feed dod to SparseMatrix; test fixes

smichr committed:
use dict=True internally

smichr committed:
do checking of Mul

smichr committed:
filter out unused coeffs

asmeurer commented:
Perhaps only somewhat related to this PR, but I found the implicit use of symbols as parameters for undetermined systems confusing (see the discussion at https://github.com/sympy/sympy/pull/23638#discussion_r911964052). Personally I still don't think this behavior should be there at all by default, but as it is, it's a little less confusing with tuple solutions than with dict solutions because at least the parameterizations are present, for example

```py
>>> solve(cos(x) - y, [x, y])
[(x, cos(x))]
>>> solve(cos(x) - y, [x, y], dict=True)
[{y: cos(x)}]
```

Maybe it would be a good idea to add parameters as trivial entries in the dict, like `{x: x, y: cos(x)}`.

smichr committed:
fix oo loop

smichr committed:
add solve_output and document match

smichr committed:
fix nits

smichr committed:
reword explanation re match

smichr committed:
remove trailing space

smichr committed:
edit output

smichr committed:
clean up post-process type wrangling

smichr commented:
@asmeurer , the set and dict output are economical while the tuple output is verbose. Redundant values can be added (if desired) like
```python
sol = list_of_dicts = solve(eqs, syms, dict=True)
for d in sol:
    for s in syms:
        ok = d.setdefault(s,s)
```
But except for the explicit visual feedback, I don't know of any value added in doing so.

smichr committed:
modify slow test

smichr commented:
@asmeurer the Explanation file has been added.

smichr committed:
fix output docstring

smichr committed:
watch for solution=None; fix output

smichr commented:
The biggest breaking change here is requiring the use of the keyword `match` when passing a single equation and many symbols. There is a difference of opinion of what `solve(Eq(a*x**2 + b*x, 3*x**2), a, b)` should mean: 1) solve for either a *or* b or 2) solve for a *and* b. And a less trivial equation perhaps points to the problem of managing `match` logic: `solve(expand((a*x+b-2*x-5)*(c-4)),a,b,c)`. If one looks for solutions in factors and applies match logic then the solutions are `[{a:2, b:5},{c:4}]`. That can be done, but it is an abuse of the situation in which one would be matching coefficients. And what would be done if the `c-4` factor were changed to `c-b`? Would that factor be skipped because a solution independent of {a,b,c} cannot be made or would an algebraic solution be returned?

smichr committed:
add not to output re {} when using match

smichr committed:
update tests

smichr committed:
supply generator for match=True if nec

smichr committed:
make factor removal more robust

asmeurer commented:
> But except for the explicit visual feedback, I don't know of any value added in doing so.

The point is to be more explicit about when a symbol is being used as a parameter. I think it's also convenient to actually include every requested symbol in the output dictionary.

smichr committed:
add tuple=True flag for solve output

smichr commented:
Since the verbose output already exists for nonlinear lists of equations, it seems least dramatic to just allow a flag to make that the default output. `tuple=True` will do this.

smichr committed:
touch-up edits

oscarbenjamin commented:
The question is not whether `solve(Eq(a*x**2 + b*x, 3*x**2), a, b)` solves for `a and b` or whether it solves for `a or b`. Clearly both a and b are listed as unknowns so it solves for both. However this is an underdetermined system so a finite set of solutions is not possible. The question then is how to represent the set of possible solutions for a and b. One way is:
```python
In [14]: solve([Eq(a*x**2 + b*x, 3*x**2)], a, b)
Out[14]: 
⎧     b    ⎫
⎨a: - ─ + 3⎬
⎩     x    ⎭
```
That's what `solve` already does for a list of of length one and there should be no distinction between `solve(eq, vars)` and `solve([eq], vars)`.

smichr commented:
The algebraic solution is now default and one must select ' match' to do otherwise....and that will disable the algebraic attempt.  

I have no further changes planned for this.

smichr commented:
OK, I don't see anything objectionable here. The `match` option must be selected (i.e. is no longer default), and raises if the system is not recognized or if the input is not a single expression. The verbose output showing values for all variables (even those who can have arbitrary values) is selectable with `tuple=True`.

oscarbenjamin commented:
Please don't merge PRs without review.

There are several things that could have been improved if properly reviewed:

1. There is already a lot of docs work adding guides on understanding various solver functions. It would have been good if the new guide added here was considered in that context and reviewed by @bertiewooster and @asmeurer.
2. The `_handle_creation_inputs` functions are already far too complex and this makes them worse. It would have been better to add a `from_dod` class method (note that I already said this on #23748).
3. There are too many different things put together in one PR here making it difficult to review. This should have been broken into separate PRs.
4. The commits should have been rewritten as there are too many tiny commits here.
5. The `solve` function is already a massive mess and this makes it more messy.
6. The `solve` signature is already a mess and this adds even more flags.
7. This is adding a new method on Basic and it's one that I already suggested a change to in #23748)

I could list more but it takes time to go through something like this. If you want your PRs to get reviewed properly then:

1. Make each PR as small as possible.
2. Clearly explain in each one what change it makes and why and what the implications are.
3. Wait more than 24 hrs.

If you are continuously pushing commits into a PR over a period of a month and building up a thousand line diff then you can't expect that others will continuously review your work in detail. It takes time to look through something like this and no one is going to study it in detail if it is not clear yet that you have even finished with it. If this was my PR I would not have merged it or even asked anyone to review it but instead split out some of the changes into separate PRs.

oscarbenjamin commented:
I see you've opened #23810 and made some changes in #23748 but I think that approach is going to make it very difficult to keep track of what has been changed or not. I think it would be much cleaner just to do a straight-forward revert of this PR and then afterwards just merge in the changes that are wanted in smaller pieces (with review).
